.github/github_workflow_scripts/purge_branch_protection_rules.py

```diff
@@ -0,0 +1,358 @@
+# Create logger
+logger = logging.getLogger()
+logger.setLevel(logging.DEBUG)  # Set the lowest level to capture all messages
+
+# Create FileHandler and set level to DEBUG
+file_handler = logging.FileHandler(f"{Path(__file__).stem}.log")
+file_handler.setLevel(logging.DEBUG)
+file_handler.setFormatter(logging.Formatter(LOG_FORMAT))
+
+# Create StreamHandler and set level to INFO
+stream_handler = logging.StreamHandler()
+stream_handler.setLevel(logging.INFO)
+stream_handler.setFormatter(logging.Formatter(LOG_FORMAT))
+
+# Add handlers to the logger
+logger.addHandler(file_handler)
+logger.addHandler(stream_handler)
+
+
+@dataclass
+class BranchProtectionRule:
+    id: str
+    pattern: str
+    matching_refs: int
+
+
+# Queries
+GET_BRANCH_PROTECTION_GRAPHQL_QUERY_TEMPLATE = """query($owner: String!, $name: String!) {
+repository(owner: $owner, name: $name) {
+    branchProtectionRules(first: 100) {
+    nodes {
+        id
+        pattern
+        matchingRefs(first: 10) {
+        totalCount
+        }
+    }
+    }
+}
+}"""
+
+DELETE_BRANCH_PROTECTION_RULE_QUERY_TEMPLATE = """mutation deleteBranchProtectionRule($branchProtectionRuleId: ID!) {
+    deleteBranchProtectionRule(input: {branchProtectionRuleId: $branchProtectionRuleId}) {
+        clientMutationId
+    }
+}"""
+
+
+# Helper Functions
+
+def get_repo_owner_and_name() -> tuple[str, str]:
+    """
+    Extracts the repository owner and name from a given repository string.
+
+    Returns:
+        A `Tuple[str, str]` containing the repository owner and name.
+
+    Raises:
+        `ValueError`: If the input string is not in the expected 'owner/repository' format.
+    """
+
+    repo = os.getenv(GH_REPO_ENV_VAR, DEFAULT_REPO)
+
+    parts = repo.split('/')
+
+    if len(parts) != 2:
+        raise ValueError("Input string must be in the format 'owner/repository'.")
+
+    owner, name = parts
+    return owner, name
+
+
+def convert_response_to_bpr(response: dict[str, Any]) -> list[BranchProtectionRule]:
+    """
+    Helper method to convert the response to a list
+    of `BranchProtectionRule`.
+
+    Arguments:
+    - `response` (``dict[str, Any]``): The response data.
+
+    Returns:
+    - a `list[BranchProtectionRule]`.
+
+    Raises:
+    - `KeyError | AttributeError` in case the conversion fails.
+    """
+
+    rules: list[BranchProtectionRule] = []
+
+    try:
+        for node in response.get('data', {}).get('repository').get('branchProtectionRules').get('nodes'):
+            rule = BranchProtectionRule(
+                id=node.get("id"),
+                pattern=node.get("pattern"),  # ignore: type
+                matching_refs=node.get("matchingRefs").get("totalCount")
+            )
+
+            rules.append(rule)
+    except (KeyError, AttributeError) as e:
+        raise e.__class__(f"{e.__class__.__name__} parsing '{response=}' as a branch protection rule: {e}")
+
+    return rules
+
+
+def should_delete_rule(rule: BranchProtectionRule) -> bool:
```

Comment:
some functions you use "rule" some "bpr" and other "branch_protection_rules" stick to one term and use it everywhere.